### ISSUE TITLE
[release/7.0] CI failure: Build Linux x64 release SourceBuild, unable to optimize assemblies for performance, runtime package not found

### ISSUE THREAD

**Issue by carlossanlop** [dotnet/runtime#80284]
Found in at least two unreleated `release/7.0` PRs.
- Both failed in the `Build Linux x64 release SourceBuild` queue.
- Both happening in a `banana.24` RID.

First PR:
- https://github.com/dotnet/runtime/pull/80250
- Job result: https://dev.azure.com/dnceng-public/public/_build/results?buildId=126826&view=logs&j=000ba8e3-34f8-51d2-8d06-9b61f10256d0&t=ca9b6e26-8c16-57f9-673e-2bc0dc655f25

Second PR:
- https://github.com/dotnet/runtime/pull/79273
- Job result: https://dev.azure.com/dnceng-public/public/_build/results?buildId=126997&view=logs&j=000ba8e3-34f8-51d2-8d06-9b61f10256d0&t=ca9b6e26-8c16-57f9-673e-2bc0dc655f25&l=97

- Output log (same in both PRs):
```
--------------------------------------------------------------------------------------
/__w/1/s/.dotnet/sdk/7.0.100/MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:/__w/1/s/artifacts/log/Release/ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:false /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=true /p:__ToolsetLocationOutputFile=/__w/1/s/artifacts/toolset/7.0.0-beta.22572.6.txt /t:__WriteToolsetLocation /warnaserror /__w/1/s/artifacts/toolset/restore.proj
/__w/1/s/.dotnet/sdk/7.0.100/MSBuild.dll /nologo -logger:/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/net7.0/Microsoft.DotNet.ArcadeLogging.dll -maxcpucount /m -verbosity:m /v:minimal /bl:/__w/1/s/artifacts/log/Release/Build.binlog /clp:Summary /nr:false /p:TreatWarningsAsErrors=true /p:ContinuousIntegrationBuild=true /p:Configuration=Release /p:RepoRoot=/__w/1/s/ /p:Restore=true /p:Build=true /p:ArcadeBuildFromSource=false /p:Rebuild=false /p:Test=false /p:Pack=true /p:IntegrationTest=false /p:PerformanceTest=false /p:Sign=false /p:Publish=true /p:TargetArchitecture=x64 /p:BuildArchitecture=x64 /p:CMakeArgs= /p:TargetRid=banana.24-x64 /p:SourceBuildNonPortable=true /p:ArcadeBuildFromSource=true /warnaserror /__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/Build.proj
  Cloning repository at: /__w/1/s/ -> /__w/1/s/artifacts/source-build/self/src/ ...
  Creating empty clone at: /__w/1/s/artifacts/source-build/self/src/
  Source repository is shallow...
  Cloning into '/__w/1/s/artifacts/source-build/self/src'...
  Checking out sources...
  Clone complete: /__w/1/s/ -> /__w/1/s/artifacts/source-build/self/src/
  __DistroRid: centos.7-x64
  /__w/1/s/.dotnet/sdk/7.0.100/MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /clp:Summary /nr:false /p:TreatWarningsAsErrors=false /p:ContinuousIntegrationBuild=false /p:Configuration=Release /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:Restore=true /p:Build=true /p:ArcadeBuildFromSource=false /p:Rebuild=false /p:Test=false /p:Pack=false /p:IntegrationTest=false /p:PerformanceTest=false /p:Sign=false /p:Publish=false /p:BuildAllConfigurations=true /p:OutputRid=banana.24-x64 /p:PortableBuild=false /p:TargetArchitecture=x64 /p:BuildArchitecture=x64 /p:CMakeArgs= /p:NoPgoOptimize=true /p:KeepNativeSymbols=true /p:RuntimeOS=banana.24 /p:ContinuousIntegrationBuild=true /p:BuildDebPackage=false /p:EnableNgenOptimization=false /p:EnablePackageValidation=false /p:DisableSourceLink=false /p:AdditionalRuntimeIdentifierParent=banana.24 /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir= /__w/1/s/artifacts/source-build/self/package-cache/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/Build.proj
    Determining projects to restore...
    Restored /__w/1/s/artifacts/source-build/self/package-cache/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/Tools.proj (in 1.35 sec).
    Determining projects to restore...
  /__w/1/s/.dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1094: Unable to optimize assemblies for performance: a valid runtime package was not found. Either set the PublishReadyToRun property to false, or use a supported runtime identifier when publishing. When targeting .NET 6 or higher, make sure to restore packages with the PublishReadyToRun property set to true. [/__w/1/s/artifacts/source-build/self/src/Build.proj]
  /__w/1/s/.dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1094: Unable to optimize assemblies for performance: a valid runtime package was not found. Either set the PublishReadyToRun property to false, or use a supported runtime identifier when publishing. When targeting .NET 6 or higher, make sure to restore packages with the PublishReadyToRun property set to true. [/__w/1/s/artifacts/source-build/self/src/Build.proj]
  
  Build FAILED.
  
  /__w/1/s/.dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1094: Unable to optimize assemblies for performance: a valid runtime package was not found. Either set the PublishReadyToRun property to false, or use a supported runtime identifier when publishing. When targeting .NET 6 or higher, make sure to restore packages with the PublishReadyToRun property set to true. [/__w/1/s/artifacts/source-build/self/src/Build.proj]
  /__w/1/s/.dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1094: Unable to optimize assemblies for performance: a valid runtime package was not found. Either set the PublishReadyToRun property to false, or use a supported runtime identifier when publishing. When targeting .NET 6 or higher, make sure to restore packages with the PublishReadyToRun property set to true. [/__w/1/s/artifacts/source-build/self/src/Build.proj]
      0 Warning(s)
      2 Error(s)
  
  Time Elapsed 00:00:32.22
  Build failed with exit code 1. Check errors above.
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/SourceBuild/SourceBuildArcadeBuild.targets(162,5): error MSB3073: The command "./build.sh  --arch x64 --configuration Release --allconfigurations --verbosity minimal --nodereuse false --warnAsError false --outputrid banana.24-x64 --portablebuild false /p:NoPgoOptimize=true /p:KeepNativeSymbols=true /p:RuntimeOS=banana.24 /p:ContinuousIntegrationBuild=true /p:BuildDebPackage=false /p:EnableNgenOptimization=false /p:EnablePackageValidation=false /p:DisableSourceLink=false /p:AdditionalRuntimeIdentifierParent=banana.24 /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir=" exited with code 1. [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/Build.proj]
##[error].packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/SourceBuild/SourceBuildArcadeBuild.targets(162,5): error MSB3073: The command "./build.sh  --arch x64 --configuration Release --allconfigurations --verbosity minimal --nodereuse false --warnAsError false --outputrid banana.24-x64 --portablebuild false /p:NoPgoOptimize=true /p:KeepNativeSymbols=true /p:RuntimeOS=banana.24 /p:ContinuousIntegrationBuild=true /p:BuildDebPackage=false /p:EnableNgenOptimization=false /p:EnablePackageValidation=false /p:DisableSourceLink=false /p:AdditionalRuntimeIdentifierParent=banana.24 /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir=" exited with code 1.

Build FAILED.

/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/SourceBuild/SourceBuildArcadeBuild.targets(162,5): error MSB3073: The command "./build.sh  --arch x64 --configuration Release --allconfigurations --verbosity minimal --nodereuse false --warnAsError false --outputrid banana.24-x64 --portablebuild false /p:NoPgoOptimize=true /p:KeepNativeSymbols=true /p:RuntimeOS=banana.24 /p:ContinuousIntegrationBuild=true /p:BuildDebPackage=false /p:EnableNgenOptimization=false /p:EnablePackageValidation=false /p:DisableSourceLink=false /p:AdditionalRuntimeIdentifierParent=banana.24 /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir=" exited with code 1. [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/7.0.0-beta.22572.6/tools/Build.proj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:01:00.49
Build failed with exit code 1. Check errors above.
```

```json
{
  "ErrorMessage": "The command \"./build.sh  --arch x64 --configuration Release --allconfigurations --verbosity minimal --nodereuse false --warnAsError false --outputrid banana.24-x64 --portablebuild false /p:NoPgoOptimize=true /p:KeepNativeSymbols=true /p:RuntimeOS=banana.24 /p:ContinuousIntegrationBuild=true /p:BuildDebPackage=false /p:EnableNgenOptimization=false /p:EnablePackageValidation=false /p:DisableSourceLink=false /p:AdditionalRuntimeIdentifierParent=banana.24 /p:ArcadeInnerBuildFromSource=true /p:DotNetBuildFromSource=true /p:RepoRoot=/__w/1/s/artifacts/source-build/self/src/ /p:ArtifactsDir=/__w/1/s/artifacts/source-build/self/src/artifacts/ /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild.binlog /p:SourceBuildOutputDir=/__w/1/s/artifacts/source-build/ /p:SourceBuiltBlobFeedDir=\" exited with code 1",
  "BuildRetry": false
}
```

**Comment by jeffschwMSFT** [comment]
@mangod9

**Comment by mangod9** [comment]
> Both happening in a banana.24 RID.

is this a dummy RID? Guessing crossgen2 is not aware of this hence PublishReadyToRun is failing with it. Is this a regression or a new build scenario?

**Comment by mangod9** [comment]
@trylek and @AntonLapounov as well. Assume there is a way to repro locally?

**Comment by AntonLapounov** [comment]
I assume this is an intermittent issue.  I could not find the relevant binlog file, maybe we should publish this artifact to help investigations:
> RESTORE_TASK_BINLOG_PARAMETERS=/__w/1/s/artifacts/source-build/self/src/artifacts/log/Release/Restore-Build.proj.binlog

**Comment by tmds** [comment]
> .dotnet/sdk/7.0.100

Could be due to missing https://github.com/dotnet/sdk/pull/29052.
This is part of 7.0.101 and later.

**Comment by AntonLapounov** [comment]
Is this still an issue?

**Comment by tmds** [comment]
https://github.com/dotnet/source-build/issues/3219#issuecomment-1461033787 was closed by using 7.0.101.
That matches my earlier comment.

So if these builds are now also using 7.0.101 or later, the issue should be gone.

**Comment by akoeplinger** [comment]
Closing since release/7.0 is using 7.0.113 and we haven't seen this issue since.
